### ISSUE TITLE
WebClient DataBufferLimitException errors

### ISSUE THREAD

**Issue by markxj** [graphql-java-generator/graphql-maven-plugin-project#131]
The WebClient created via the SpringConfiguration uses the default buffer limit of ~256K. Performing a query that returns more data than that causes the following exception

```
org.springframework.web.reactive.function.client.WebClientResponseException: 200 OK from POST http://localhost:8080/graphql; nested exception is org.springframework.core.io.buffer.DataBufferLimitException: Exceeded limit on max bytes to buffer : 262144
        at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:229) ~[spring-webflux-5.3.13.jar:5.3.13]
        at org.springframework.web.reactive.function.client.DefaultClientResponse.lambda$createException$1(DefaultClientResponse.java:207) ~[spring-webflux-5.3.13.jar:5.3.13]

```

This can be worked around by manually creating the WebClient, but a long term fix would be to change the client_spring_configuration.vm.java template

```
	@Bean
	@ConditionalOnMissingBean(name = "webClient${springBeanSuffix}")
	public WebClient webClient${springBeanSuffix}(String graphqlEndpoint${springBeanSuffix}, //
			@Autowired(required = false) @Qualifier("httpClient${springBeanSuffix}") HttpClient httpClient${springBeanSuffix},
			@Autowired(required = false) @Qualifier("serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}",
                         **CodecCustomizer defaultCodecCustomizer,**
) ServerOAuth2AuthorizedClientExchangeFilterFunction serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}) {
		return GraphQLConfiguration.getWebClient(graphqlEndpoint${springBeanSuffix}, httpClient${springBeanSuffix}, serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}, defaultCodecCustomizer);
	}
```

And the method in GraphQLConfiguration to

```
public static WebClient getWebClient(String graphqlEndpoint, HttpClient httpClient, **CodecCustomizer codecCustomizer,**
			ExchangeFilterFunction... filters) {
		Builder webClientBuilder = WebClient.builder()//
				.baseUrl(graphqlEndpoint)//
				.defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
				.defaultUriVariables(Collections.singletonMap("url", graphqlEndpoint))
				.**codecs(configurer -> codecCustomizer.customize(configurer)**);

		if (httpClient != null) {
			webClientBuilder.clientConnector(new ReactorClientHttpConnector(httpClient));
		}
		if (filters != null) {
			for (ExchangeFilterFunction filter : filters) {
				if (filter != null) {
					webClientBuilder.filter(filter);
				}
			}
		}

		return webClientBuilder.build();
	}	
```
This then allows the webclient to be configured using the spring.codec.* properties - e.g. spring.codec.max-in-memory-size.

It might also be worth looking at a method that doesn't block when returning data but instead returns a stream that the client can consume.

**Comment by etienne-sf** [comment]
Hello,

Thanks for the proposal. 

But I'm not convinced. 

There are so many hook point that could be added, everywhere, that the plugin's use would be really hard to understand (and to document, actually). 

I think that overriding the WebClient allows all changes in it. 
Whereas add specific configuration parameters would need plenty of entry points. 


Was it that complex to override the WebClient ?

Étienne

**Comment by etienne-sf** [comment]
Hello,

Thanks for the proposal. 

But I'm not convinced. 

There are so many hook points that could be added, everywhere, that the plugin's use would be really hard to understand (and to document, actually). 

On the contrary, I think that overriding the WebClient allows all changes that could be needed. 

Was it that complex to override the WebClient ?

Étienne

**Comment by markxj** [comment]
It wasn't hard to do once I figured out that the generator wasn't using the recommended spring application properties for configuring the WebClient

**Comment by etienne-sf** [comment]
Ok

So your point here is to better document this limitation of the default WebClient implementation by the plug-in, right ?

Or do you mean that, if I update the plug-in's code as you proposed,then the parameter from the application.properties file would be automatically taken into account ?

**Comment by etienne-sf** [comment]
I did a try with the code update of your initial post.

I have two points here:
* The code you proposed generates an execution time error (see below)
* From the searched I did on the net, it seems that this `spring.codec.max-in-memory-size` parameter doesn't work. And you have to manually call the `webClientBuilder.codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(XXX))` method


I tried the code you proposed, in particular:

```java
	@Bean
	@ConditionalOnMissingBean(name = "webClient${springBeanSuffix}")
	public WebClient webClient${springBeanSuffix}(String graphqlEndpoint${springBeanSuffix}, //
			@Autowired(required = false) @Qualifier("httpClient${springBeanSuffix}") HttpClient httpClient${springBeanSuffix},
			@Autowired(required = false) @Qualifier("serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}",
                         CodecCustomizer defaultCodecCustomizer
) ServerOAuth2AuthorizedClientExchangeFilterFunction serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}) {
		return GraphQLConfiguration.getWebClient(graphqlEndpoint${springBeanSuffix}, httpClient${springBeanSuffix}, serverOAuth2AuthorizedClientExchangeFilterFunction${springBeanSuffix}, defaultCodecCustomizer);
	}
```

When it initializes, Spring complains about the missing `CodecCustomizer` bean: it find no bean that can be set to the `defaultCodecCustomizer` parameter. I could make it, but I think it would hide other Spring configuration parameters.

To sum up: there seems to be a bug in Spring Boot, which makes that you have to manually call the `maxInMemorySize` method.

So the workaround you used seems to be the only way to do that.

**Comment by markxj** [comment]
My workaround was basically this:

```
   @Bean
    public WebClient webClientTest(String graphqlEndpointTest,
                                          CodecCustomizer defaultCodecCustomizer,
                                          @Qualifier("serverOAuth2AuthorizedClientExchangeFilterFunctionTest") ServerOAuth2AuthorizedClientExchangeFilterFunction serverOAuth2AuthorizedClientExchangeFilterFunctionTest)
    {
        return WebClient.builder()
                .baseUrl(graphqlEndpointTest)
                .defaultHeader("Content-Type", "application/json")
                .defaultUriVariables(Collections.singletonMap("url", graphqlEndpointTest))
                .filter(serverOAuth2AuthorizedClientExchangeFilterFunctionTest)
                .codecs(configurer -> codecCustomizer.customize(configurer))
                .build();
    }
```
So it is using the bean for defaultCodeCustomizer - this is with spring boot 2.4.13 - not sure which version you were building against

**Comment by etienne-sf** [comment]
Right !

I changed the parameter name, when I tried it. 

Setting it back to ´defaultCodeCustomizer´ makes it work. 

Very nice !

I was waiting the end of this discussion to build a new release. 
So I'll release it this week end.

**Comment by etienne-sf** [comment]
Now available in the 1.18.5 release
